### PR TITLE
Refactor, optimise and update the tube network example

### PR DESCRIPTION
**Pull request by sorsaffari**
### Main objectives

This PR is aimed to achieve the following goals in the Tube Network example:
1. Have it synced with the latest Grakn Core (master)
2. Refactor the code for better maintainability and understandability 
3. Optimise the migration script to be an example of using _generic template functions_ and _concurrent transactions_.
4. Optimise loading of data for the purpose of rendering the visualiser

### Changes made per source file

**`src/migration.py`**: the migration script takes the following steps:
1. while reading the `.json` files:
    -  creates the `input` (with a pre-defined structure) to be used by the template functions
    -  passes the structured `input` on to corresponding template function (`entity_template(data)` or `relationship_template(data)` and in return gets the constructed Graql insert query
    -  validates the returned query for uniqueness (via generic and custom validators), before appending it the dictionary that holds all queries
2. flattens the dictionary containing all queries into a list to be used for concurrent insertion
3. concurrently executes the queries meant to create instances of entities
4. concurrently executes the queries meant to create instances of relationships

**`src/statistics.py`**: 
1. enters the interactive mode in the terminal upon execution
2. the user then selects a question among the presented list or chooses to get the answers to all questions
3. selection of any questions displays the question, the query used to retrieve the answer and the answer itself.

**`src/journey_planner.py`**: takes the same approach as `statistics.py`.
1. asks the user for the name of the origin station
2. verifies the existence of the given station by its name 
3. asks the user for the name of the destination station
4. verifies the existence of the given station by its name 
5. asks the user to select the shortest path strategy (fewest stops or route changes)
6. displays the name of the stations located on the shortest path

**`src/app.py`**: besides general refactoring, the special improvement was made by using one single expressive query to retrieve all the data required for rendering the interface.

### Future changes
1. `src/app.py` can be further refactored to abstract away non-Graql related parts of the code. this helps the users to focus on how Grakn powers the application.
2. with the upcoming changes in Client Python and Grakn Core, there may be the need for another round of updates to keep in sync with 1.5.

### CODE REVIEW

File: applications/tube_network/README.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+```bash
+cd path-to-cloned-repository/applications/tube_network
+python3 -m tube_network.src.migration
+```
+
+The content of [`src/migration.py`](src/migration.py) is a python script that:
+1. as it goes through the TFL's `.json` files, constructs dictionaries with a pre-defined structure that get passed on to the template functions for constructing Graql relationship/entity insert queries.
+2. the constructed Graql insert queries, after basic uniqueness validation, get stored as items of arrays within a dictionary.
+3. the dictionary containing all the Graql queries, gets flattened, to prepare the data in two chunks of entities and relationships for a series of concurrent insertions.
```
Comment (jmsfltchr) [inline applications/tube_network/README.md]: the dictionary containing all the Graql queries is flattened to prepare the data in two chunks, one chunk of entities and one chunk of relationships, for a series of concurrent insertions.

File: applications/tube_network/README.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+4. lastly, a set of processes initiate the set of transactions that perform the Graql insert queries on the `tube_network` keyspace.
+
+Once complete, you have stored the tube network data in Grakn!
+Now you're ready to start playing with the data.
+
+
+## Reasoning
```
Comment (jmsfltchr) [inline applications/tube_network/README.md]: This isn't actually reasoning is it! Maybe eat title should be _insights_

File: applications/tube_network/README.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+Now you can query the database, you can try asking more complex questions, like asking for the time between stations of over 8 minutes:
+```
+match
+    $sta1 isa station, has name $sta1-nam;
+    $sta2 isa station, has name $sta2-nam;
+    $sec isa route-section, has duration $dur; $dur > 8;
+    (beginning: $sta1, end: $sta2, service: $sex) isa tunnel;
```
Comment (jmsfltchr) [inline applications/tube_network/README.md]: `service: $sex` - plus this demure variable isn't even necessary ;)

File: applications/tube_network/README.md (hunk 2 above)
Comment (jmsfltchr) [reply applications/tube_network/README.md]: If we want to give a rule example, we can use a rule to determine whether a train is eastbound or westbound

File: applications/tube_network/README.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+Opening the demo app for the first time may take up to a minute.
+
+### The Basics
+
+- Drag mouse to pan
+- Press `=`/`+` to zoom in
+- Press `-`/`_` to zoom out
+- Press `c` to clear the map of analytics
+
+#### Find Shortest Path
```
Comment (jmsfltchr) [inline applications/tube_network/README.md]: We should add images of using the demo app! Both for shortest path and for analytics

File: applications/tube_network/src/journey_planner.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                else:
+                    from_station_id = from_station_list[0].map().get("sta").id
+                    valid_from_name = True
+
+        # Get the destination station
+        print("")
+        valid_to_name = False
+        while not valid_to_name:
+            to_station_name = input("Where to: ")
+            print("Verifying station bane '" + to_station_name + "' ...")
```
Comment (jmsfltchr) [inline applications/tube_network/src/journey_planner.py]: _station name_ not _station bane_?

File: applications/tube_network/src/statistics.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+import tube_network.settings as settings
+
+def print_to_log(title, content):
+  print(title)
+  print("")
+  print(content)
+  print("\n")
+
+# How many stations do exist?
+def execute_query_1(question, transaction):
```
Comment (jmsfltchr) [inline applications/tube_network/src/statistics.py]: These functions should be named as per the comments, to describe their functions, maybe prefix them all with _query_ to indicate they're part of a family

File: applications/tube_network/src/statistics.py [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+    print_to_log("Query:", query)
+
+    answer = list(transaction.query(query))[0]
+    median_duration = answer.number()
+
+    print("Median of durations: " + str(median_duration))
+
+
+# What's the median duration among all trips?
+def execute_query_7(question, transaction):
```
Comment (jmsfltchr) [inline applications/tube_network/src/statistics.py]: All of these execute_query functions except 3 and 4 repeat themselves, so they could all be simplified by extracting a method

File: applications/tube_network/src/statistics.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+    ]
+
+    print_to_log("Query:", "\n".join(query))
+    query = "".join(query)
+
+    answers = transaction.query(query).collect_concepts()
+    result = [ answer.value() for answer in answers ]
+
+
+    print_to_log("West most stations with " + str(lat) + " are: ", result)
```
Comment (jmsfltchr) [inline applications/tube_network/src/statistics.py]: In fact minimum latitude will determine the northernmost stations! Latitude goes north-south, longitude goes east-west

File: applications/tube_network/README.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    $sta2 isa station, has name $sta2-nam;
+    $sec isa route-section, has duration $dur; $dur > 8;
+    (beginning: $sta1, end: $sta2, service: $sex) isa tunnel;
+    (route-operator: $tul, section: $sec) isa route;
+    $tul isa tube-line, has name $tul-nam;
+get $sta1-nam, $sta2-nam, $tul-nam, $dur; limit 30;
+```
+
+## Retrieve Statistical Information
+To gain an overall understanding of the available statistical queries in Grakn, try the [`src/statistics.py`](src/statistics.py).
```
Comment (jmsfltchr) [inline applications/tube_network/README.md:66]: The UI for the statistics looks much more mature now, perhaps some docs on using it, or a screenshot of some cool terminal interaction would be good

File: applications/tube_network/README.md (hunk 2 above)
Comment (sorsaffari) [reply applications/tube_network/README.md]: good point! changing the title _Insights_ and created an issue for adding a rule to the schema (https://github.com/graknlabs/examples/issues/49)

File: applications/tube_network/src/statistics.py (hunk 7 above)
Comment (sorsaffari) [reply applications/tube_network/src/statistics.py]: @jmsfltchr we definitely should code clean when we're building products. when it comes to examples though, the objective is to showcase Grakn's features, Graql and usage of Clients. In this case, the query functions show how 1) a question is written in Graql 2) a Graql query gets executed, and 3) answers are retrieved. extracting any part of this code into another method breaks this flow

File: applications/tube_network/src/statistics.py (hunk 6 above)
Comment (sorsaffari) [reply applications/tube_network/src/statistics.py]: very good point!

**Pull request merged**